staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -1523,6 +1546,10 @@ type Result struct {
 	err         error
 	statusCode  int
 
+	// Log calls in Result methods use the same context for logging as the
+	// method which created the Result. This context has no cancellation.
+	loggingCtx context.Context
```

Comment:
The `loggingCtx` is needed for:
```
	out, _, err := r.decoder.Decode(r.body, &schema.GroupVersionKind{Version: "v1"}, nil)
	if err != nil {
		ctx := r.loggingCtx
		if ctx == nil {
			ctx = context.Background()
		}
		klog.FromContext(ctx).V(5).Info("Body was not decodable (unable to check for Status)", "err", err)
		return r.err
	}
```

I agree that this isn't particularly likely, but consider if we don't convert this and you are trying to debug an issue that occurs for one particular request by grepping for some unique key/value pair that was set for all activities related to that request: this log entry here might explain the issue, but you won't see it.

I could extract the logger unconditionally when creating the request. That then incurs the overhead for `FromContext` whether this error occurs or not. OTOH, `WithoutCancel` also isn't free. Without good benchmarks it's hard to tell which one is cheaper, and both are probably cheap compared to the cost of handling a request.

Yet another alternative is to store a `getLogger` function in the `Request` instead of the context. Instead of `WithoutCancel`, the field would get initialized with `getLogger: func() klog.Logger { klog.FromContext(ctx} }`. But that then binds the context for longer.

I'm leaning towards storing the logger - changed in a separate commit for review and to document the decision.